### ISSUE TITLE
usb4tos-1-19-cur not working right ...

### ISSUE THREAD

**Issue by franklukas** [freemint/freemint#218]
Test usb4tos-1-19-cur on my Standard Atari Falcon with a Netusbee. I works not right. When I copy a folder from a USB stick (500MB BGM) the error message comes up "Ungüldige Operation" in german afterwards no more access to the usb stick and I must reboot to have again access to the USB stick.

When I copy many (8) files the error message comes up "Fehler beim Lesen einer Datei".

The old Netusbee Beta05 works right.

**Comment by Perdrix24** [comment]
Do these [drivers](https://af81b87b-6cdc-4e20-a03a-8c2d6e44cbd0.filesusr.com/archives/3a09df_9a7cc058fc6546ac9f0ec7cfa4f82fc1.zip?dn=netusb06.zip) work?

Please see the readme file, see if any of the issues at the bottom of the file apply to your problem.

**Comment by franklukas** [comment]
Try netusb06 with the Pakete folder from EasyMiNT (14MB-87MB 10 files). It works fine. My Atari Falcon is stock without any (Clock) patches and has 4MB Ram + Magnum Falcon Alternate Ram card so I have 10MB Ram in total.

**Comment by Perdrix24** [comment]
What OS are you using (TOS, MagiC, EmuTOS) and what version of OS?

**Comment by franklukas** [comment]
TOS 4.04

**Comment by Perdrix24** [comment]
Try replacing storage.prg in usb4tos-1-19-cur with the one from NETUSB06, using all the other drivers from usb4tos-1-19-cur to see if any difference.

**Comment by franklukas** [comment]
It works with storeage.prg from NETUSB06 but not 100% perfect. I copy 6 files, 3 little ones and 3 big ones (40MB). 5 files are by done but them shows not up on the USB stick folder. The folder was empty, after I do a reboot 5 from the 6 files show up.

**Comment by franklukas** [comment]
I test the whole usb4tos again and the errors comes up when I copy files from the usb stick to the hdd. Copy from hdd to the usb stick are ok I think.

**Comment by Perdrix24** [comment]
The usb4tos-1-19 drivers work fine for me, just like the NETUSB06 drivers. It passed my infamous 1,000 files, 37MB copy, including a 9MB file, to and from the USB drive.

Try this: Rename USBTOOL.ACC to USBTOOLX.ACC . This will cause polling to be suspended while the accessory window is closed. The polling is known to cause problems in rare cases. So close the window before copying your files. 

I'm working on USBTOOL to see if I can fix this problem.

**Comment by franklukas** [comment]
It work not for me. I use the NETUSB05 again.

**Comment by troed** [comment]
I just installed usb4tos-1-19-cur on my Unicorn equipped STE, and while it finds the Unicorn root hub it finds no USB devices. Previously I used some much earlier version (sorry, don't know which at the moment) and all was well, even several USB hubs linked.

I don't know if it would be the same issue, but "not working right" fits ...

**Comment by mikrosk** [comment]
@troed can you try https://www.perdrixapps.com/usb (v07 at the time if writing) ?

Also, v05 can be found here: https://lotharek.pl/files/netusbee/NETUSB05.ZIP does that work for you?

**Comment by troed** [comment]
Hi Miro,

There's no UNICORN.PRG in those packages, but what I did was to replace USB.PRG and STORAGE.PRG and keep the rest from the 1-19 package. Let me know if that was a bad idea.

The result was that NETUSB07 behaved exactly like 1-19, and NETUSB06 at least found something​ - either a hub or displaying just a dot or mostly just immediately resetting the computer when scanning.

I will try NETUSB05 next, but it's a different API version so then I need to downgrade the other programs too.

regards,
Troed

------- Original Message -------

**Comment by mikrosk** [comment]
I guess @Perdrix24 and @DavidGZ are more capable of answering that but unless the PRG files complain to each other, it should be fine. What about other peripherals? Do they work fine in the latest version? (i.e. USB sticks) If so, then it's really interesting, i.e. if it finds an USB stick but not the USB network card; my understanding is that USB bus scanning is the UCD device's responsibility (i.e. NETUSBEE.PRG) so it should give consistent results.

**Comment by troed** [comment]
Hi,

So the latest drivers I've previously used for the Unicorn are from 2018-01-01 - and everything works well with those. It seems there's an API step already in between those and NETUSB05 though so I couldn't test with it either (I have one UNICORN.PRG from 2018 and one from the latest USB4TOS and nothing inbetween).

My setup is a built in Unicorn in an STE, and it's usually only used with a USB stick (thus UNICORN.PRG is what scans the bus). The reason I started meddling with this today was that I have also connected a USB-Ethernet adapter (previously tested working) and wanted to add ETH.PRG. That's when I ended up with version mishaps.

When UNICORN.PRG (2018 version) scans the bus it lists the Unicorn Root Hub, a USB 2.0 Hub, the USB memorystick and the USB-Ethernet adapter. If I upgrade to the latest USB4TOS (1-19) then UNICORN.PRG (2022 version) finds nothing except the Unicorn Root Hub when scanning. Keeping that UNICORN.PRG and downgrading USB.PRG / STORAGE.PRG to the ones in NETUSB07 results in the same behaviour. Using the ones in NETUSB06 means "something" happens when UNICORN.PRG scans the bus, but it's clearly broken data and bugs out.

So I can't be certain the problem lies with USB.PRG rather than UNICORN.PRG in the latest USB4TOS. I'd be happy to bisect a bit to figure out when this started - I have vague memories of building these drivers many years ago when I debugged a Unicorn device.

regards,
Troed

------- Original Message -------

**Comment by DavidGZ** [comment]
> I guess @Perdrix24 and @DavidGZ are more capable of answering that but unless the PRG files complain to each other, it should be fine. What about other peripherals? Do they work fine in the latest version? (i.e. USB sticks) If so, then it's really 

There have been a couple of USB API changes since 2018, during the driver's loading the drivers should have complained about an API mismatch if he has mixed drivers from 2018 and recent ones.

**Comment by DavidGZ** [comment]
I have taken a look to the commits in the Unicorn driver since 2018, I've found some suspicious ones. Specially one from February 2018 where there was change in the delay function for 68000 machines.

**Comment by DavidGZ** [comment]
>So I can't be certain the problem lies with USB.PRG rather than UNICORN.PRG in the latest USB4TOS. I'd be happy to bisect a bit to figure out when this started - I have vague memories of building these drivers many years ago when I debugged a Unicorn device. regards, Troed

Besides to the fact that I don't have too much time to look into this I don't have the hardware (Unicorn), so if you could bisect this yourself it would help a lot.

**Comment by DavidGZ** [comment]
> > I guess @Perdrix24 and @DavidGZ are more capable of answering that but unless the PRG files complain to each other, it should be fine. What about other peripherals? Do they work fine in the latest version? (i.e. USB sticks) If so, then it's really
> 
> There have been a couple of USB API changes since 2018, during the driver's loading the drivers should have complained about an API mismatch if he has mixed drivers from 2018 and recent ones.

Wrong, it seems only MiNT modules check the API versions, TOS drivers don't do it :-/

**Comment by mikrosk** [comment]
> Wrong, it seems only MiNT modules check the API versions, TOS drivers don't do it :-/

Actually they do check something because I had an USB.PRG which refused to work with latest NETUSBEE.PRG or so.

**Comment by troed** [comment]
Hi all,

Right, I guess I have some information now.

Non-working
2020-01-02
2019-01-06
2018-07-14
2018-04-22
2018-02-09 make[5]: *** No rule to make target '../../../tosdelay.c', needed by 'sl811-hcd.o'. Stop.

Working
2018-01-01 (? that's the _compile_ date on the latest ones from alanh I have. I believe they were downloaded from freemint.org at the time)
Basically no version works until I get back to what looks to be like a little merge chaos in the beginning of 2018, where it no longer compiles due to the lack of tosdelay.c. When sorting on commit date, things also start to go a bit out of order at that time with commits from 2018 and 2017 intermingled in commit history.

All the tests have been done in the same way. Compile, put the binaries on my boot drive and see whether the Unicorn is able to detect any device on the USB bus when booting.

regards,
Troed

------- Original Message -------

**Comment by mikrosk** [comment]
Hey Troed, thanks for taking time to bisect it. I have prepared more builds for you (btw all of them are compiling nicely but sometimes you have to do make clean to make it work): [usb.zip](https://github.com/freemint/freemint/files/8237445/usb.zip)

2018-02-08 is 868bedb9, I think there's a solid chance that this is the one which broke your setup because there are several delays added. Before that the last chance happend on Oct 2017, so you are right, the January build you have is just a build from January, not a date of some change.

**Comment by troed** [comment]
Hi Miro,

Thanks - your 2018-04-02 works. I do do make clean, I'll look at why I'm not getting tcpdelay.c to appear when checking out those older commits then.

Next suspect is probably;

79ed9148 2018-04-16 Add real delay functions to TOS USB drivers

I'll look at that one now.

regards,
Troed

------- Original Message -------

**Comment by troed** [comment]
Alright, that was indeed the one.

After the commit below the Unicorn is no longer finding anything when scanning the bus. Also, while it takes a good second or so to find the first hub with working drivers, now it seems to skip to not having found anything very quickly.

* 79ed9148 2018-04-16 Add real delay functions to TOS USB drivers

https://github.com/freemint/freemint/commit/79ed914837283283939d450136557db3afb6522d

I'll be happy to test any suggested changes.

regards,
Troed
------- Original Message -------

**Comment by DavidGZ** [comment]
OK, thanks a lot for bisecting. I think the problem is the that we don't calibrate the loops count for the delay function in the Unicorn driver. Let me compile a new driver for you to test.

**Comment by DavidGZ** [comment]
@troed , I've attached a new Unicorn driver for testing.

[unicorn.zip](https://github.com/freemint/freemint/files/8237977/unicorn.zip)

**Comment by troed** [comment]
Test successful :D Much appreciated!!!

regards,
Troed

------- Original Message -------

**Comment by DavidGZ** [comment]
Great! Thanks! I'll commit the fix.

I leave the issue open, because the original problem for Frank is unrelated with this fix.

**Comment by DavidGZ** [comment]
@franklukas there has been some changes in the NetUSBee driver. Could you please test again with the latest snapshot driver?

Also we've come to the conclusion that power going through the ROMport is to unreliable for the NetUSBee's USB function, we're encouraging people to connect their USB devices through a powered hub. If you have accessed to one you could try to see if it helps with your NetUSBee issues.

**Comment by mikrosk** [comment]
> Also we've come to the conclusion that power going through the ROMport is to unreliable for the NetUSBee's USB function

I would maybe rephrase it that you should check whether the power delivered on ROM port is +5V as the first thing. It seems that mine is around 3V due to [blown fuse](https://www.atari-forum.com/viewtopic.php?p=423279#p423279) and it makes NetUSBee unreliable _in general_, not only when it comes to powering USB peripherals.

**Comment by DavidGZ** [comment]
> 
> I would maybe rephrase it that you should check whether the power delivered on ROM port is +5V as the first thing. It seems that mine is around 3V due to [blown fuse](https://www.atari-forum.com/viewtopic.php?p=423279#p423279) and it makes NetUSBee unreliable _in general_, not only when it comes to powering USB peripherals.
> 

Even if the power is +5V there must be some kind unreliability in the quality of the power going through the NetUSBee. We have always needed delays in the read/write routines for any hardware but the ST line, and suddenly all those delays aren't needed thanks to a powered hub, you've experienced this with your CT60 too. This is why I suspect putting the 68000 NetUSBee driver (without delays) in v7 driver's  package may be was a not a good idea because I think Claude tested the drivers with a powered hub and he thought that they work well with the 030 and 060 CPU too, but this was not the case before powered hubs were used.

Also note that a successful test was not a simple USB device detection and moving some KB from/to a USB storage device, we transfer 100 MB in both directions, sometimes when everything seemed OK the transfer failed after some minutes, this was a failed test. Another reason why debugging this driver has been nightmare, tests took minutes.

**Comment by mikrosk** [comment]
> Even if the power is +5V there must be some kind unreliability in the quality of the power going through the NetUSBee. We have always needed delays in the read/write routines for any hardware but the ST line, and suddenly all those delays aren't needed thanks to a powered hub, you've experienced this with your CT60 too. This is why I suspect putting the 68000 NetUSBee driver (without delays) in v7 driver's package may be was a not a good idea because I think Claude tested the drivers with a powered hub and he thought that they work well with the 030 and 060 CPU too, but this was not the case before powered hubs were used.

No no, this is not true. During the evolution ;) of #258 it became clear that the only malfunctioning thing was my TT's ROM port. CT60 always worked reliably (after putting CACHE_ON.PRG in AUTO), I'm not using the hub neither on CT60 nor on my MegaSTE. So the delays are really not needed.

As soon as I replace the fuse I'll do more tests without the hub on my TT to confirm whether the second keyboard from #261 is going to be detected (in that case I am right and we don't need any delays) or not (in that case you are right and the hub is not recommended but _essential_ to use).

**Comment by DavidGZ** [comment]
> No no, this is not true. During the evolution ;) of #258 it became clear that the only malfunctioning thing was my TT's ROM port. CT60 always worked reliably (after putting CACHE_ON.PRG in AUTO), I'm not using the hub neither on CT60 nor on Falcon030 nor on MegaSTE. So the delays are really not needed there.
> 

Oops....then I misunderstood the "evolution" of the thread :-)....sorry.

In any case there is something I don't get, It's a fact that we needed delays with the 030-060 CPUs, and suddenly they're not needed anymore. What happened? The change in the compiler version?

I thought that the unreliable power source theory fit perfectly with what we have experienced, USB transfers randomly breaking, sometimes after some seconds sometimes after several minutes, I thought that random power drops could explain this behavior. What I don't see in this scenario is why the delays would have been helped.

**Comment by DavidGZ** [comment]
> No no, this is not true. During the evolution ;) of #258 it became clear that the only malfunctioning thing was my TT's ROM port. CT60 always worked reliably (after putting CACHE_ON.PRG in AUTO), I'm not using the hub neither on CT60 nor on Falcon030 nor on MegaSTE. So the delays are really not needed there.
> 

Oops....then I misunderstood the "evolution" of the thread :-)....sorry.

In any case there is something I don't get, It's a fact that we needed delays with the 030-060 CPUs, and suddenly they're not needed anymore. What happened? The change in the compiler version?

I thought that the unreliable power source theory fit perfectly with what we have experienced, USB transfers randomly breaking, sometimes after some seconds sometimes after several minutes, I thought that random power drops could explain this behavior. What I don't see in this scenario is why the delays would have been helped.

**Comment by mikrosk** [comment]
> What happened? The change in the compiler version?

That's a good question. As we know, NETUSB06 was compiled on @Perdrix24's gcc 4.6.4 while NETUSB07 was taken from a USB4TOS snapshot compiled by gcc 7.5.0.  However from Claude's replies I got the feeling that he regularly tests his builds on Falcon030/060 so I'd assume that even NETUS000.PRG from NETUSB06 worked well on his Falcon (which he implied by the statement that he stopped including NETUSBEE.PRG for that very reason).

When I manage to fix my TT's ROM port, I'll try to make another round of testing with gcc 4.6.4, too. It's true that at first it _seemed_ that your commit 7d81217e helped on the TT but with 3V voltage levels the whole NetUSBee must be ruled out as a totally unreliable source of truth.

**Comment by mikrosk** [comment]
I was able to reproduce it on my TT + USB hub + latest USB4TOS + `NETUS000.PRG` exactly as described in https://github.com/freemint/freemint/issues/218#issue-829419485. For me the issue was present even when copying _to_ USB stick, not _from_ it as Frank described.

That's good news.

And you know what is another good news? `NETUSBEE.PRG` with @DavidGZ's changes did fix it!

So that explains why it worked for Frank in NETUSB05 (compiled by gcc4 = slower code) and not in USB4TOS (compiled by gcc7 = faster code) as discussed in https://github.com/freemint/freemint/issues/258#issuecomment-1166512740.

Great job @DavidGZ in persuading me to do these tests! Closing as fixed by https://github.com/freemint/freemint/commit/7d81217e45bfe1b569285cd47d1cc60e6ba9cf04. @franklukas please download latest USB4TOS and definitely use `NETUSBEE.PRG` and not  `NETUS000.PRG` on your Falcon. If the issue persists, feel free to reopen it.

**Comment by DavidGZ** [comment]
> I was able to reproduce it on my TT + USB hub + latest USB4TOS + `NETUS000.PRG` exactly as described in [#218 (comment)](https://github.com/freemint/freemint/issues/218#issue-829419485). For me the issue was present even when copying _to_ USB stick, not _from_ it as Frank described.

@mikrosk that USB hub was powered, wasn't it?

**Comment by mikrosk** [comment]
@DavidGZ Yep.
